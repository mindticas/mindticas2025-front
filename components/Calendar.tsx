'use client';

import { useState, useMemo, useEffect } from 'react';
import {
    Badge,
    Box,
    Button,
    Flex,
    Grid,
    GridItem,
    Heading,
    useBreakpointValue,
} from '@chakra-ui/react';
import { useBookingContext } from '@/context/BookingContext';
import { DateTime } from 'luxon';
import { getAppointments } from '@/services/AppointmentService';
import { format } from 'date-fns';
import { es } from 'date-fns/locale';

// Available time slots
const timeSlots = [
    '10:00',
    '11:00',
    '12:00',
    '13:00',
    '14:00',
    '16:00',
    '17:00',
];

// Function to check if a date is Sunday
const isSunday = (date: Date) => date.getDay() === 0;

export default function Calendar() {
    const { setDateTime, dateTime } = useBookingContext();
    const [error, setError] = useState<string | null>(null);

    // Check if the screen is small (responsive)
    const isSmallScreen = useBreakpointValue({ base: true, sm: false });

    // State for selected date and time
    const [selectedDate, setSelectedDate] = useState<Date | null>(null);
    const [selectedTime, setSelectedTime] = useState<string | null>(null);
    const [bookedTimes, setBookedTimes] = useState<string[]>([]);

    // Function to get the current month in Spanish
    const getMonth = () =>
        format(selectedDate ?? new Date(), 'MMMM', { locale: es });

    // fetch booked appointments
    useEffect(() => {
        const fetchAppointments = async () => {
            try {
                const data = await getAppointments();
                const bookedTimes = data.map((appointment) => {
                    return appointment.scheduled_start;
                });
                setBookedTimes(bookedTimes);
            } catch (error) {
                setError(
                    'No se pueden cargar los horarios ocupados. Inténtalo de nuevo más tarde.',
                );
            }
        };

        // if dateTime is reset to null, reset selectedDate and selectedTime
        if (!dateTime) {
            setSelectedDate(null);
            setSelectedTime(null);
        }

        fetchAppointments();
    }, [dateTime]);

    // Generate dates for the current week (13 days: 7 days before, today, and 5 days after)
    const weekDates = useMemo(() => {
        const today = new Date();
        const startOfWeek = new Date(today);
        startOfWeek.setDate(today.getDate() - today.getDay()); // Adjust to the start of the week (Sunday)

        return Array.from({ length: 13 }, (_, i) => {
            const date = new Date(startOfWeek);
            date.setDate(startOfWeek.getDate() + i); // Add consecutive days
            return date;
        });
    }, []);

    // Handle date selection
    const handleDateSelect = (date: Date) => {
        if (!isSunday(date)) {
            // Disallow selecting Sundays
            setSelectedDate(date);
            setSelectedTime(null); // Reset selected time
        }
    };

    // Handle time selection
    const handleTimeSelect = (time: string) => {
        setSelectedTime(time);
        if (selectedDate) {
            const [hours, minutes] = time.split(':');
            // Combine date and time in format 'YYYY-MM-DDTHH:MM:SS' with luxon
            const dateTime = DateTime.fromObject(
                {
                    year: selectedDate.getFullYear(),
                    month: selectedDate.getMonth() + 1,
                    day: selectedDate.getDate(),
                    hour: parseInt(hours, 10),
                    minute: parseInt(minutes, 10),
                },
                { zone: 'utc' },
            );

            const isoString = dateTime.toISO();
            // Set the selected date and time in the BookingContext
            setDateTime(isoString);
        }
    };

    // Check if a date is today
    const isToday = (date: Date) => {
        const today = new Date();
        return date.toDateString() === today.toDateString();
    };

    // Check if a date is in the past (before today)
    const isPastDate = (date: Date) => {
        const today = new Date();
        return date < new Date(today.setHours(0, 0, 0, 0)); // Compare with the start of the current day
    };

    // Check if a time is in the past (only applies if the selected date is today)
    const isPastTime = (time: string) => {
        if (!selectedDate || !isToday(selectedDate)) return false; // Only applies if the date is today
        const [hours, minutes] = time.split(':');
        const timeDate = new Date(selectedDate);
        timeDate.setHours(
            Number.parseInt(hours, 10),
            Number.parseInt(minutes, 10),
        ); // Combine date and time
        return timeDate < new Date(); // Compare with the current time
    };

    const bookedTimesForDate = useMemo(() => {
        if (!selectedDate) return [];

        // Format the selected date in ISO format (without time)
        const selectedDateISO = DateTime.fromJSDate(selectedDate).toISODate();

        // Filter booked appointments for the selected date
        return bookedTimes
            .filter((bookedTime) => {
                const bookedDateISO = DateTime.fromISO(bookedTime).toISODate();
                return bookedDateISO === selectedDateISO;
            })
            .map((bookedTime) =>
                DateTime.fromISO(bookedTime, {
                    zone: 'utc',
                }).toFormat('HH:mm'),
            );
    }, [selectedDate, bookedTimes]);

    const isBookedTime = (time: string) => {
        return bookedTimesForDate.includes(time);
    };

    return (
        <>
            <div>{error && <p style={{ color: 'red' }}>{error}</p>}</div>
            <Flex direction='column' align='center' mx='auto'>
                <Box
                    className='calendar-box'
                    bg='white'
                    p={isSmallScreen ? 4 : 6}
                    borderRadius='lg'
                    boxShadow='md'
                    borderWidth='1px'
                    borderColor='gray.200'
                >
                    {/* Calendar title */}
                    <Flex justifyContent='space-between' mb={4} px={2}>
                        <Heading
                            as='h2'
                            fontSize='lg'
                            fontWeight='bold'
                            color='black'
                        >
                            Selecciona el día
                        </Heading>
                        <Badge
                            style={{ textTransform: 'capitalize' }}
                            variant='solid'
                            p={2}
                            backgroundColor='black'
                            color='white'
                        >
                            {getMonth()}
                        </Badge>
                    </Flex>

                    {/* Days of the week */}
                    <Grid
                        templateColumns='repeat(7, 1fr)'
                        gap={isSmallScreen ? 1 : 2}
                        mb={2}
                    >
                        {['Do', 'Lu', 'Ma', 'Mi', 'Ju', 'Vi', 'Sa'].map(
                            (day) => (
                                <GridItem
                                    key={day}
                                    textAlign='center'
                                    fontWeight='medium'
                                    color='gray.500'
                                >
                                    {day}
                                </GridItem>
                            ),
                        )}
                    </Grid>

                    {/* Dates of the week */}
                    <Grid
                        templateColumns='repeat(7, 1fr)'
                        gap={isSmallScreen ? 1 : 3}
                        mb={4}
                    >
                        {weekDates.map((date) => (
                            <Button
                                key={date.toISOString()}
                                p={2}
                                borderRadius='md'
                                transition='background-color 0.2s'
                                bg={
                                    isPastDate(date) || isSunday(date)
                                        ? 'gray.200' // Past dates or Sundays
                                        : selectedDate?.toDateString() ===
                                          date.toDateString()
                                        ? 'black' // Selected date
                                        : isToday(date)
                                        ? 'gray.200' // Today
                                        : 'gray.100' // Future dates
                                }
                                color={
                                    isPastDate(date) || isSunday(date)
                                        ? 'gray.400' // Past dates or Sundays
                                        : selectedDate?.toDateString() ===
                                          date.toDateString()
                                        ? 'white' // Selected date
                                        : 'black' // Future dates or today
                                }
                                _hover={
                                    isPastDate(date) || isSunday(date)
                                        ? {} // No hover for past dates or Sundays
                                        : { bg: 'gray.200' } // Hover for future dates
                                }
                                onClick={() => handleDateSelect(date)}
                                disabled={isPastDate(date) || isSunday(date)} // Disable past dates or Sundays
                            >
                                {date.getDate()} {/* Day of the month */}
                            </Button>
                        ))}
                    </Grid>

                    {/* Time selection (only if a date is selected) */}
                    {selectedDate && (
                        <Box>
                            <Heading
                                as='h3'
                                fontSize='lg'
                                fontWeight='semibold'
                                mb={2}
                                color='black'
                            >
                                Selecciona la hora
                            </Heading>
                            <Grid templateColumns='repeat(3, 1fr)' gap={2}>
                                {timeSlots.map((time) => (
                                    <Button
                                        key={time}
                                        p={2}
                                        borderRadius='md'
                                        transition='background-color 0.2s'
                                        bg={
                                            isPastTime(time) ||
                                            isBookedTime(time)
                                                ? 'gray.200' // Past times
                                                : selectedTime === time
                                                ? 'black' // Selected time
                                                : 'gray.100' // Future times
                                        }
                                        color={
                                            isPastTime(time) ||
                                            isBookedTime(time)
                                                ? 'gray.400' // Past times
                                                : selectedTime === time
                                                ? 'white' // Selected time
                                                : 'black' // Future times
                                        }
                                        _hover={
                                            isPastTime(time) ||
                                            isBookedTime(time)
                                                ? {} // No hover for past times
                                                : { bg: 'gray.200' } // Hover for future times
                                        }
                                        onClick={
                                            () =>
                                                !isPastTime(time) &&
                                                !isBookedTime(time) &&
                                                handleTimeSelect(time) // Handle time selection
                                        }
                                        disabled={
                                            isPastTime(time) ||
                                            isBookedTime(time)
                                        } // Disable past times
                                    >
                                        {time}
                                    </Button>
                                ))}
                            </Grid>
                        </Box>
                    )}
                </Box>
            </Flex>
        </>
    );
}
