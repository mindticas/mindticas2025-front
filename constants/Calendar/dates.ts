export const DAYS_OF_WEEK_ES = [
    'Domingo',
    'Lunes',
    'Martes',
    'Miércoles',
    'Jueves',
    'Viernes',
    'Sábado',
]

export const DAYS_OF_WEEK_SHORT_ES = ['Do', 'Lu', 'Ma', 'Mi', 'Ju', 'Vi', 'Sa']