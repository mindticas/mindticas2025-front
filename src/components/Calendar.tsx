import { Box, Button } from '@chakra-ui/react'
import React from 'react'
import { useColorMode } from './ui/color-mode'

export default function Calendar() {
    const { colorMode, toggleColorMode } = useColorMode()
  return (
    <Box p="16">
      <p>The current Theme is: {colorMode} mode.</p>
      <Box h="7" />
      <Button onClick={toggleColorMode}>Change Theme</Button>
    </Box>
  )
}
