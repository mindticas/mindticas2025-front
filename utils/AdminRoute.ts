import Layout from '@/components/Layout';
import { Box } from '@chakra-ui/react';

export class AdminRoute {
    // Fixed route for admin
    private static route: string = '/admin';

    // Returns the Layout component if the pathname contains '/admin', otherwise returns box
    static getLayout(pathname: string) {
        return pathname.includes(this.route) ? Layout : Box;
    }
}
