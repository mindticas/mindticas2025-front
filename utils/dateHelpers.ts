// Gets the name of the day according to the full calendar format
export const getDayName = (dateStr: string): string => {
    const [year, month, day] = dateStr.split('-').map(Number)
    const date = new Date(year, month -1, day)
    return date.toLocaleDateString('es-ES', { weekday: 'long'})
}
// Get the milliseconds in a day
export function addDays(date: Date, days: number): Date {
    const milisecondsPerDay = 24 * 60 * 60 * 1000; 
    return new Date(date.getTime() + days * milisecondsPerDay)
}