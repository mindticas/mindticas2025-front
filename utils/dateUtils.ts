import { DateTime } from "luxon";

interface DateTimeParts {
    date: string;
    time: string;
}

// * @returns Object with (YYYY-MM-DD) y time (HH:MM)

export const splitDateTimeFromISO = (
    isoString: string | null | undefined,
): DateTimeParts => {
    if (!isoString) return { date: '', time: '' };

    const dateObj = new Date(isoString);

    if (isNaN(dateObj.getTime())) {
        return { date: '', time: '' };
    }

    // format date (YYYY-MM-DD)
    const date = DateTime.fromJSDate(dateObj)
       .setZone('local')
        .toFormat('yyyy-MM-dd');

    // format time (HH:MM)
    const time = dateObj
        .toLocaleTimeString('es-ES', {
            hour: '2-digit',
            minute: '2-digit',
            hour12: false,
        })
        .replace(/\./g, '');

    return { date, time };
};
