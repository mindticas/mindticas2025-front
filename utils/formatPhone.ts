export default function formatPhone(phone: string): string {
    // Remove everything that is not digits
    const digits = phone.replace(/\D/g, '');
    // Applies format only if it has 10 digits
    if (digits.length === 10) {
        return `${digits.slice(0, 3)}-${digits.slice(3, 6)}-${digits.slice(6)}`;
    }
    return phone; // Returns the original if it doesn't match
}
