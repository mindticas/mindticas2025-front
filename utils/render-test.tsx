import { Provider } from '@/components/ui/provider';
import { BookingProvider } from '@/context/BookingContext';
import { render as rtlRender } from '@testing-library/react';

export function render(ui: React.ReactNode) {
    return rtlRender(<>{ui}</>, {
        wrapper: (props: React.PropsWithChildren) => (
            <Provider>
                <BookingProvider>{props.children}</BookingProvider>
            </Provider>
        ),
    });
}
