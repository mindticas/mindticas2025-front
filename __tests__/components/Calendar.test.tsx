import Calendar from '@/components/Calendar';
import { render } from '@/utils/render-test';
import { screen } from '@testing-library/react';

describe('Calendar', () => {
    it('renders the calendar with the current day', () => {
        render(<Calendar />);
        const today = new Date().getDate().toString();
        expect(screen.getByText(today)).toBeInTheDocument();
    });
});
