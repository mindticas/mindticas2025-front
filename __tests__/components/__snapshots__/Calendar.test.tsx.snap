// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Snapshot should match the snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,s,n,m,o,l,u,h)=&gt;{let a=document.documentElement,v=["light","dark"];function p(c){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&l?o.map(f=&gt;l[f]||f):o;k?(a.classList.remove(...S),a.classList.add(c)):a.setAttribute(y,c)}),C(c)}function C(c){h&&v.includes(c)&&(a.style.colorScheme=c)}function d(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(m)p(m);else try{let c=localStorage.getItem(s)||n,y=u&&c==="system"?d():c;p(y)}catch(c){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div />
  <div
    class="css-qj8tuy"
  >
    <div
      class="calendar-box css-vk4h8o"
    >
      <div
        class="css-kg7qdy"
      >
        <h2
          class="chakra-heading css-wu1q6v"
        >
          Selecciona el d√≠a
        </h2>
        <span
          class="chakra-badge css-1xfmew8"
          style="text-transform: capitalize;"
        >
          abril
        </span>
      </div>
      <div
        class="css-56tv3v"
      >
        <div
          class="css-1121gur"
        >
          Do
        </div>
        <div
          class="css-1121gur"
        >
          Lu
        </div>
        <div
          class="css-1121gur"
        >
          Ma
        </div>
        <div
          class="css-1121gur"
        >
          Mi
        </div>
        <div
          class="css-1121gur"
        >
          Ju
        </div>
        <div
          class="css-1121gur"
        >
          Vi
        </div>
        <div
          class="css-1121gur"
        >
          Sa
        </div>
      </div>
      <div
        class="css-16nmiu8"
      >
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          13 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          14 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          15 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          16 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          17 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          18 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          19 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          20 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          21 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          22 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          23 
        </button>
        <button
          class="chakra-button css-r87ip0"
          disabled=""
          type="button"
        >
          24 
        </button>
        <button
          class="chakra-button css-1732jk6"
          disabled=""
          type="button"
        >
          25 
        </button>
        <button
          class="chakra-button css-1732jk6"
          disabled=""
          type="button"
        >
          26 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          27 
        </button>
        <button
          class="chakra-button css-1732jk6"
          disabled=""
          type="button"
        >
          28 
        </button>
        <button
          class="chakra-button css-1732jk6"
          disabled=""
          type="button"
        >
          29 
        </button>
        <button
          class="chakra-button css-1732jk6"
          disabled=""
          type="button"
        >
          30 
        </button>
        <button
          class="chakra-button css-1732jk6"
          disabled=""
          type="button"
        >
          1 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          2 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          3 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          4 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          5 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          6 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          7 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          8 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          9 
        </button>
        <button
          class="chakra-button css-qa2o2z"
          disabled=""
          type="button"
        >
          10 
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
